# 1. Import Libraries

In [128]:
#Importing the necessary libraries
import pandas as pd

# Display of all columns in the notebook
pd.set_option("display.max_columns", None)


# 2. Load Dataset

In [129]:
# Loading the data into the Notebook
df = pd.read_csv(r"C:\Users\PC\Desktop\Future Interns\Telco_Customer_Churn.csv")

# First 5 rows of the dataset
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


# 3. Initial Data Inspection

In [130]:
# Checking the total number of rwas and columns
df.shape

(7043, 21)

Insight: The data has 21 columns (variables) and 7043 rows (entries)

In [131]:
#Inspecting the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


Insight: The dataset has different data types such as integer, float, and object.

In [132]:
# Checking null entries
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

Insight: No null values (entries) in the data 

In [133]:
# Checking duplicated rows
df.duplicated().any()


False

Insight: No duplicated entries.

In [134]:
# Descriptive statistics
print(df.describe())

       SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000


In [135]:
# Chekcing the unique entries in each column
print(df.nunique())

customerID          7043
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1585
TotalCharges        6531
Churn                  2
dtype: int64


# 4. Data Cleaning

## Fixing the data types

In [136]:
# Converting TotalCharges and MonthlyCharges to numeric data types

df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')

df['MonthlyCharges'] = pd.to_numeric(df['MonthlyCharges'], errors='coerce')

In [137]:
replace_cols = [
    'MultipleLines','OnlineSecurity','OnlineBackup',
    'DeviceProtection','TechSupport','StreamingTV','StreamingMovies'
]

for col in replace_cols:
    df[col] = df[col].replace({
        'No internet service':'No',
        'No phone service':'No'
    })

# 5. FUTURE ENGINEEERING

In [138]:
# 1. Flag for customers lacking key services (for Top 5 Churn Reasons chart)

df['NoTechSupport'] = (df['TechSupport'] == 'No').astype(int) 

df['NoOnlineSecurity'] = (df['OnlineSecurity'] == 'No').astype(int) 

df['MonthToMonth'] = (df['Contract'] == 'Month-to-month').astype(int) 

# 2. High monthly charge flag (over $80)
df['HighMonthlyCharge'] = (df['MonthlyCharges'] > 80).astype(int)

# 3. Electronic check flag
df['ElectronicCheck'] = (df['PaymentMethod'] == 'Electronic check').astype(int)

In [139]:
# Mapping the values 1s and 0s in the SeniorCitizen column to Yes and No respectively.
Mapping = {1:"Yes", 0:"No"}

df["SeniorCitizen"] = df["SeniorCitizen"].map(Mapping)

In [140]:
# Mapping churn
df['Churn_Flag'] = df['Churn'].map({'Yes': 1, 'No': 0}) 

# 5. Churn flag (0/1 for calculations)
df['Churn_Flag'] = (df['Churn'] == 'Yes').astype(int)

In [141]:
# Creating a column for high risk customers
df['HighRiskSegment'] = (
    (df['Contract'] == 'Month-to-month') & 
    (df['TechSupport'] == 'No') & 
    (df['MonthlyCharges'] > 70)
).astype(int)

Mapping = {1:"Yes", 0:"No"}

df["HighRiskSegment"] = df["HighRiskSegment"].map(Mapping)

In [142]:
df['TenureGroup'] = pd.cut(
    df['tenure'],
    bins=[0,12,24,48,60,72],
    labels=['0-12','12-24','24-48','48-60','60-72']
)

In [143]:
# Monthly charge segmentation
df['ChargeLevel'] = pd.cut(
    df['MonthlyCharges'],
    bins=3,
    labels=['Low','Medium','High']
)

In [144]:
# Total Services available to customers\
services = [
    'PhoneService','OnlineSecurity','OnlineBackup',
    'DeviceProtection','TechSupport','StreamingTV','StreamingMovies'
]

df['TotalServices'] = (df[services] == 'Yes').sum(axis=1)

# 6. Validation of the final dataset

In [145]:
print(df.describe())
print(df['Churn'].value_counts())
print(df['TenureGroup'].value_counts())

            tenure  MonthlyCharges  TotalCharges  NoTechSupport  \
count  7043.000000     7043.000000   7032.000000    7043.000000   
mean     32.371149       64.761692   2283.300441       0.709783   
std      24.559481       30.090047   2266.771362       0.453895   
min       0.000000       18.250000     18.800000       0.000000   
25%       9.000000       35.500000    401.450000       0.000000   
50%      29.000000       70.350000   1397.475000       1.000000   
75%      55.000000       89.850000   3794.737500       1.000000   
max      72.000000      118.750000   8684.800000       1.000000   

       NoOnlineSecurity  MonthToMonth  HighMonthlyCharge  ElectronicCheck  \
count       7043.000000   7043.000000        7043.000000      7043.000000   
mean           0.713332      0.550192           0.378532         0.335794   
std            0.452237      0.497510           0.485056         0.472301   
min            0.000000      0.000000           0.000000         0.000000   
25%        

# 7. Final Cleaned Data

In [146]:
# Inspecting the final cleaned dataset
print("Final Shape:", df.shape)
df.head()

Final Shape: (7043, 31)


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,NoTechSupport,NoOnlineSecurity,MonthToMonth,HighMonthlyCharge,ElectronicCheck,Churn_Flag,HighRiskSegment,TenureGroup,ChargeLevel,TotalServices
0,7590-VHVEG,Female,No,Yes,No,1,No,No,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No,1,1,1,0,1,0,No,0-12,Low,1
1,5575-GNVDE,Male,No,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.50,No,1,0,0,0,0,0,No,24-48,Medium,3
2,3668-QPYBK,Male,No,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,0,1,0,0,1,No,0-12,Medium,3
3,7795-CFOCW,Male,No,No,No,45,No,No,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No,0,0,0,0,0,0,No,24-48,Low,3
4,9237-HQITU,Female,No,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,1,1,0,1,1,Yes,0-12,Medium,1


# 8. Exporting the dataset

In [147]:
# Exporting the final cleaned data to excel
telco_cleaned = df.copy()
telco_cleaned.to_csv(
    r"C:\Users\PC\Desktop\Future Interns\FUTURE_DS_02\telco_cleaned.csv",
    index=False
)